pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+	}
+	// No need to update if ResourceVersion is not changed
+	if newApp.ResourceVersion == oldApp.ResourceVersion {
+		log.Logger.Info("No need to update because app object is not modified")
+		return
+	}
+
+	// handle spec change
+	if !equality.Semantic.DeepEqual(oldApp.Spec, newApp.Spec) {
+		//TODO: handle update, discuss what should be updated since only the queue is the common property yet
+	}
+}
+
+/*
+Remove the application from the scheduler as well
+*/
+func (appMgr *AppManager) deleteApp(obj interface{}) {
+	app, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	appID := app.Name
+	err := appMgr.amProtocol.RemoveApplication(appID)
+	if err != nil {
+		log.Logger.Error("Application removal failed",
+			zap.String("appID", appID),
+			zap.Error(err))
+	}
+	log.Logger.Debug("App CRD deleted",
+		zap.String("Name", appID))
+	//TODO: delete related pods and deployments
+}
+
+/*
+Add application to scheduler
+*/
+func (appMgr *AppManager) addApp(obj interface{}) {
+	appCRD, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	if appMeta, ok := appMgr.getAppMetadata(appCRD); ok {
+		app := appMgr.amProtocol.GetApplication(appMeta.ApplicationID)
+		if app == nil {
+			appMgr.amProtocol.AddApplication(&interfaces.AddApplicationRequest{
+				Metadata: appMeta,
+				//TODO: debug recovery
+				//TODO: try to define the owner reference for the submitted pods to this CRD and try of we delete the
+				//CRD the pods will be deleted as well
+				Recovery: false,
+			})
+			// set and save status = New
+			appMgr.updateAppCRDStatus(appCRD, appv1.NewApplicationState)
+		}
+	}
+}
+
+func (appMgr *AppManager) updateAppCRDStatus(appCRD *appv1.Application, status appv1.ApplicationStateType) {
+	copy := appCRD.DeepCopy()
+	copy.Status = appv1.ApplicationStatus{
+		AppStatus:  status,
+		Message:    "app CRD status change",
+		LastUpdate: v1.NewTime(time.Now()),
+	}
+	_, err := appMgr.controller.appClient.ApacheV1alpha1().Applications(appCRD.Namespace).UpdateStatus(copy)
+	if err != nil {
+		log.Logger.Error("Failed to update application CRD",
+			zap.String("AppId", copy.Name))
+		return
+	}
+}
+
+func (appMgr *AppManager) getAppMetadata(app *appv1.Application) (interfaces.ApplicationMetadata, bool) {
+	appID := app.Name
+
+	// tags will at least have namespace info
+	// labels or annotations from the pod can be added when needed
+	// user info is retrieved via service account
+	tags := map[string]string{}
+	if app.Namespace == "" {
+		tags[constants.AppTagNamespace] = constants.DefaultAppNamespace
+	} else {
+		tags[constants.AppTagNamespace] = app.Namespace
+	}
+
+	return interfaces.ApplicationMetadata{
+		ApplicationID: appID,
+		QueueName:     app.Spec.Queue,
+		User:          "default",
+		Tags:          tags,
+	}, true
+}
+
+func convertShimAppStateToAppCRDState(status string) appv1.ApplicationStateType {
+	switch status {
+	case corecache.New.String():
```

Comment:
we should not depend on the core side package directly,
based on current limitations, I think we will just do the parse with strings...

Reply:
removed